File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.java
class Solution {
    public int longestOnes(int[] nums, int k) {
        
        /*
        nums = [1,1,1,0,0,0,1,1,1,1,0], k = 2
                s 
                  i 
                start = 0 
                maxOnes = 2
                count = 0 
        
        */
        
        int start = 0;
        int maxOnes = 0;
        int count = 0; 
        
        for(int i = 0; i< nums.length; i++) {
            
            if (nums[i] == 0 && count <= k){
                count += 1;
                
            
            }
            
            while(count > k){
                if(nums[start] == 0){
                    count -=1;
                }
                start += 1;
            }
            
            
            
            
            maxOnes = Math.max(maxOnes, (i - start) + 1);
            
            
        }
        
        
        return maxOnes;
        
    }
}

File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        
        '''
        [0,0,1,1,0,0,1,1,1,0,1,1,0,0,0,1,1,1,1] k = 3

           s 
                   e 
                count = 3
                maxOnes = 5
                start = 0 
                end = 0 
                
                
                
        sliding window or two pointer 
        
        
        
        
        
        '''
        count = 0 
        maxOnes = 0 
        start = 0 
        
        for end in range(len(nums)): 
            if nums[end] == 0 and count <=k:
                count += 1
            
                
                
            while count > k: 
                if nums[start] == 0:
                    count -= 1 
                start += 1
                    
            maxOnes = max(maxOnes, end - start + 1)
            
                
                    
            
            
            
        return maxOnes
                
        